packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -622,18 +622,24 @@ final RegExp _kDeviceRegex = new RegExp(r'^(\S+)\s+(\S+)(.*)');
 /// [mockAdbOutput] is public for testing.
 List<AndroidDevice> getAdbDevices({ String mockAdbOutput }) {
   List<AndroidDevice> devices = <AndroidDevice>[];
-  List<String> output;
+  String text;
 
   if (mockAdbOutput == null) {
     String adbPath = getAdbPath(androidSdk);
     if (adbPath == null)
       return <AndroidDevice>[];
-    output = runSync(<String>[adbPath, 'devices', '-l']).trim().split('\n');
+    text = runSync(<String>[adbPath, 'devices', '-l']);
   } else {
-    output = mockAdbOutput.trim().split('\n');
+    text = mockAdbOutput;
   }
 
-  for (String line in output) {
+  // Check for error messages from adb
+  if (!text.contains('List of devices')) {
+    printError(text);
```

Comment:
That was my initial thought as well and I considered throwing an exception. Looking at call sites for this method, I realized it's not really an exception, but rather additional information for the user. This method is used by higher level functions to gather information about devices, and this method (and others) are already calling `printError` as a sidecar way to provide additional information to the user.

That said, I agree that `printError` here (and elsewhere throughout this method and others) is not ideal in that it prevents calling methods from choosing how to present this additional information to the user.